### DISCUSSION TITLE
Contents of Assets ("Resources")

### DISCUSSION THREAD

**Discussion by baboulebou** [openid/authzen#24] [General]
What properties should the Asset object  have?

Proposing that the ID should be mandatory.

**Comment by ogazitt** [comment]
I believe this could be completely symmetrical with what I was proposing for Subject (aka "Principal"):

```
{
  "type": "some-object-type",
  "identifier": "some-object-id",
  "properties": {
    "foo": "bar"
  }
}
```

where `properties` is optional.

**Comment by tulshi** [comment]
In a response, one could possibly have an asset description that is only a class of assets, and not a specific asset (see issue #31 ). This is why the `id` field is left optional

**Comment by tr33** [comment]
I think there are two perspectives to that:
a) how is a resource/subject referred to in a request?
b) how is a resource/subject referred to in a response

In the most simple case, a PEP just might want to put the whole object in the request to PDP to evaluation and give the PDP full object context for evaluation - and thats it. (like a reference-by-copy)
In other scenarios, the PEP+PDP or PDP only are connected to a PIP and only refer to objects in that PIP (reference-by-id). Using just an “id” here might be a bit mono-dimentional to refer to arbitrary objects. Most RDBs/GraphDBs/ReBAC have some sort of "type" for objects. Plus, a object type would become a necessary criteria if the PDP has to chose between objects in multiple PIPs (i.e. one for articles, one for prices, ...)

So for the REQUEST case, I see the need for both scenarios:

    reference by copy: give the PDP the complete resource/subject in JSON notation, however complex it is
    reference by id: give the PDP and object-type and object-id which refer to an object in a 3rd party resource (PIP or PDPs internal store)

Both scenarios should be should be differentiated and covered in the message format of the request part.
